src/requirements/reqs-functions.ts

```diff
@@ -206,46 +207,57 @@ function iterateThroughCollegeOrMajorRequirements(
 
-    const { name: requirementName, courses } = requirement;
+    const { name: requirementName, courses: requirementCourses } = requirement;
 
     let totalRequirementCredits = 0;
     let totalRequirementCount = 0;
-    const coursesThatFulilledRequirement: string[] = [];
+    const coursesThatFulfilledRequirement: CourseTaken[][] = requirementCourses.map(() => []);
+    const courseCodesThatFulfilledRequirement = new Set<string>();
 
     // eslint-disable-next-line no-loop-func
     coursesTaken.forEach(courseTaken => {
-      const {
-        code, roster, subject, number, credits
-      } = courseTaken;
-      if (courses[roster] && courses[roster][subject] && courses[roster][subject].includes(number)) {
-        // depending on what it is fulfilled by, either increase the count or credits you took
-        switch (requirement.fulfilledBy) {
-          case 'courses':
-            totalRequirementCount += 1;
-            break;
-          case 'credits':
-            totalRequirementCredits += credits;
-            break;
-          case 'self-check':
-            return;
-          default:
-            throw new Error('Fulfillment type unknown.');
+      const { roster, subject, number } = courseTaken;
+      requirementCourses.forEach((subRequirementCourses, subRequirementIndex) => {
+        if (subRequirementCourses[roster] && subRequirementCourses[roster][subject] && subRequirementCourses[roster][subject].includes(number)) {
+          // add the course to the list of courses used to fulfill that one sub-requirement
+          coursesThatFulfilledRequirement[subRequirementIndex].push(courseTaken);
         }
+      });
+    });
+    // eslint-disable-next-line no-loop-func
+    coursesThatFulfilledRequirement.forEach((coursesThatFulfilledSubRequirement, subRequirementIndex) => {
+      if (coursesThatFulfilledSubRequirement.length === 0) {
+        return;
+      }
+      // depending on what it is fulfilled by, either increase the count or credits you took
+      switch (requirement.fulfilledBy) {
+        case 'courses':
+          totalRequirementCount += 1;
+          break;
+        case 'credits':
+          totalRequirementCredits += coursesThatFulfilledSubRequirement
+            .map(course => course.credits)
+            .reduce((a, b) => Math.max(a, b), 0);
+          break;
+        case 'self-check':
+          return;
+        default:
+          throw new Error('Fulfillment type unknown.');
+      }
 
-        // add the course to the list of courses used to fulfill that one requirement
-        coursesThatFulilledRequirement.push(code);
-
+      coursesThatFulfilledSubRequirement.forEach(({ code }) => {
         // Add course to dictionary with name
         if (code in satisfiedRequirementMap) satisfiedRequirementMap[code].push(requirementName);
         else satisfiedRequirementMap[code] = [requirementName];
-      }
+        courseCodesThatFulfilledRequirement.add(code);
+      });
```

Comment:
@hanywang2 @willespencer 
For each sub-requirement, we can potentially put multiple courses into the requirement fulfillment map.
This is not a problem with my implementation: the old implementation also implicitly does this (After all, the old one can't even distinguish between sub-requirements 😞).
I'm not sure how the `satisfiedRequirementMap` is used elsewhere, so I will leave the question to you.

Reply:
Can you give an example? I'm having trouble understanding

Reply:
Consider the CS core requirement:
```json
[
  ["CS 2800", "CS 2802"],
  ["CS 3110"],
  ["CS 3410", "CS 3420"],
  ["CS 4410"],
  ["CS 4820"]
]
```

Suppose I have `["CS 2800", "CS 2802"]`, then the `satisfiedRequirementMap`  will contain
```json
{
  "CS2800": ["CS Core"],
  "CS2802": ["CS Core"],
}
```

I'm not exactly sure about the purpose of `satisfiedRequirementMap`, so I'm not sure whether the setup will cause problems.

Reply:
The `satisfiedRequirementMap` is used for the information icon on the course card. It's not on display right now, but there should be a small information icon that when you hover over it on the course, it will display the requirement that it satisfies. Most ideally, we would want it to display like this if you have `["CS 2800", "CS 2802", "CS 3110", "CS 3410", "CS 3420"]`
```
{
   "CS 2800": [{type: "info": requirement: "CS Core"}]
   "CS 2802": [{type: "duplicate": requirement: "CS Core"}]
   "CS 3110": [{type: "info": requirement: "CS Core"}]
   "CS 3410": [{type: "info": requirement: "CS Core"}]
   "CS 3420": [{type: "duplicate": requirement: "CS Core"}]
}
```
...or something along these lines. We haven't implemented the icons yet so they could be subject to change, but anything that is informative to the user on which requirement all courses satisfy (and ideally sub-requirements tho the previous implementation doesn't address that)

Reply:
We would want an implementation for `satisfiedRequirementMap` for this branch because we will be needing it very soon